lib/ae_mdw_web/controllers/block_controller.ex

```diff
@@ -18,7 +19,27 @@ defmodule AeMdwWeb.BlockController do
   Endpoint for block info by hash or kbi.
   """
   @spec block(Conn.t(), map()) :: Conn.t()
-  def block(%Conn{assigns: %{state: state}} = conn, %{"hash_or_kbi" => hash_or_kbi} = params) do
+  def block(%Conn{assigns: %{state: state}} = conn, %{"hash_or_kbi" => hash_or_kbi}) do
+    with {:ok, kbi} <- WebUtil.parse_gen(state, hash_or_kbi),
+         {_prev_cursor, [block | _], _next_cursor} <-
```

Comment:
Perhaps we can avoid using State functions in the controller and leave them up to the contexts

That way we can also avoid checking for it twice

```elixir
case Validate.non_neg(hash_or_kbi) do
  {:ok, kbi} -> 
    case Blocks.fetch_blocks(kbi..kbi, 1,..) do
      {nil, [block], nil} ->
        json(conn, block)
      {nil, [], nil} ->
        {:error, ErrInput.NotFound.exception(value: hash_or_kbi)}
    end
  {:error, _reason} -> 
    with {:ok, block} <- Blocks.fetch(state, hash_or_kbi) do
      json(conn, block)
    end
 end
```

Reply:
I think with `State.exists?(state, Model.Block, {gen, -1})` it makes more explicit the validation of the input but I don't mind changing to this similar code.